### PR TITLE
Kernel/HID: Add support for virtio input devices (+ use them for RISC-V in run.py)

### PR DESCRIPTION
**Pull request by spholz**
This PR makes keyboard and (absolute) mouse input work on RISC-V!

The QEMU virtio keyboard, mouse and tablet (absolute mouse) work on both x86-64 and riscv64.
I only tested the keyboard driver with my german keyboard (+german layout), but that should hopefully not be an issue. I would still appreciate it if someone could test this with a different layout.

We could also consider changing run.py to use the virtio input devices on x86-64 in the future, since those might be a better option than using PS/2 when running in QEMU.

The new virtio driver is not in the Bus/VirtIO directory, as #24152 will move all other virtio drivers out of there.

### CODE REVIEW

File: Kernel/Devices/HID/VirtIO/EvDevDefinitions.h
Comment (spholz) [inline Kernel/Devices/HID/VirtIO/EvDevDefinitions.h:1]: This file is pretty big (ELFABI.h has a similar line count though), but I don't know if I should strip it down. We don't need all definitions in there, but we might perhaps need some more in the future.

File: Kernel/Devices/HID/VirtIO/Input.cpp
Comment (spholz) [inline Kernel/Devices/HID/VirtIO/Input.cpp:212]: With #24210 merged, we could remove this devid matching code and instead always create a MouseDevice and KeyboardDevice for each virtio input device, which might be preferable?

File: Kernel/Devices/HID/VirtIO/Input.cpp
Comment (Hendiadyoin1) [inline Kernel/Devices/HID/VirtIO/Input.cpp:61]: Can you also assert the size of the inner structs?

File: Kernel/Bus/VirtIO/Transport/PCIe/Detect.cpp
Comment (ADKaster) [inline Kernel/Bus/VirtIO/Transport/PCIe/Detect.cpp:47]: :( Would be nice to refactor this so that some device or device manager or PCI Bus Master or something owned these objects instead of leak_ref-ing them.

File: Kernel/Devices/HID/VirtIO/EvDevDefinitions.h
Comment (ADKaster) [reply Kernel/Devices/HID/VirtIO/EvDevDefinitions.h:1]: Seems like we're accumulating these definition files from third parties :) This one seems to be from FreeBSD though, so meh. Seems fine to just keep the file as-is.

File: Kernel/Devices/HID/VirtIO/Input.cpp
Comment (ADKaster) [reply Kernel/Devices/HID/VirtIO/Input.cpp:209]: :shrug: Seems like a reasonable refactor after that architecture PR is ready to rock n roll.

File: Kernel/Arch/riscv64/PCI/Initializer.cpp
Comment (ADKaster) [inline Kernel/Arch/riscv64/PCI/Initializer.cpp:196]: worth an issue to track this problem?

**Review by ADKaster — COMMENTED** [review]
Couple comments, but overall I don't see any actual problems with the code.

File: Kernel/Bus/VirtIO/Transport/PCIe/Detect.cpp
Comment (spholz) [reply Kernel/Bus/VirtIO/Transport/PCIe/Detect.cpp:47]: #23448 should make this a lot cleaner.

File: Kernel/Arch/riscv64/PCI/Initializer.cpp
Comment (spholz) [reply Kernel/Arch/riscv64/PCI/Initializer.cpp:196]: #22720 is about the same underlying issue. The only difference is that SeaBIOS allocates BARs above 4G in that issue, not us.

File: Kernel/Devices/HID/VirtIO/EvDevDefinitions.h
Comment (spholz) [reply Kernel/Devices/HID/VirtIO/EvDevDefinitions.h:1]: I used this FreeBSD header since I wasn't sure how/if I legally could make our own definitions file, since it would use the same numbers (and probably similar names).
I obviously can't use the linux evdev header, since it's GPL.

**Comment by supercomputer7** [comment]
> We could also consider changing run.py to use the virtio input devices on x86-64 in the future, since those might be a better option than using PS/2 when running in QEMU.

If the VirtIO driver is stable enough, I'd suggest that we should move to that and ditch PS/2 for bare metal usage.

File: Kernel/Devices/HID/VirtIO/Input.cpp
Comment (spholz) [inline Kernel/Devices/HID/VirtIO/Input.cpp:233]: There sadly is no nice way to ignore unknown devids, since the caller of `initialize_virtio_resources` uses `MUST.` We also would have to leave the Input device in a half-initialized state, which doesn't seem ideal.
So I would rather wait for #24210 to be merged, since I don't think we should be using `VERIFY_NOT_REACHED` here.

Comment (spholz) [reply Kernel/Devices/HID/VirtIO/Input.cpp]: I removed the devid matching code now.
Only the mouse will reliably work until #24210 is merged.

The keyboard actually works with the way it is added in run.py because it is listed first, and QEMU apparently assigns PCI addresses based on the order of the command line arguments.
As a result, the keyboard is located at "/dev/input/keyboard/0", which is where WindowServer expects it to be.

**Comment by ADKaster** [comment]
conflicts

**Comment by spholz** [comment]
rebased

**Review by ADKaster — APPROVED** [review]

**Pull request merged**
